dashboard/lib/pd/survey_pipeline/survey_rollup_decorator.rb

```diff
@@ -0,0 +1,135 @@
+module Pd::SurveyPipeline
+  class SurveyRollupDecorator < SurveyPipelineWorker
+    def self.process_data(context)
+      check_required_input_keys self::INPUT_KEYS, context
+
+      results = decorate(*context.values_at(*self::INPUT_KEYS))
+
+      self::OUTPUT_KEYS.each do |key|
+        context[key] ||= {}
+        context[key].deep_merge! results[key]
+      end
+
+      context
+    end
+
+    def self.decorate(summaries, facilitator_id, current_workshop_id, related_workshop_ids,
+      submissions, submission_type, parsed_questions, question_categories, errors = [])
+      result = {
+        facilitators: {},
+        current_workshop: current_workshop_id,
+        related_workshops: {},
+        facilitator_response_counts: {this_workshop: {}, all_my_workshops: {}},
+        facilitator_averages: {},
+        errors: errors
+      }
+
+      # Populate result[:facilitators] = {fac_id => fac_name}
+      fac_name = User.find(facilitator_id)&.name || "UserId_#{facilitator_id}"
+      result[:facilitators][facilitator_id] = fac_name
+
+      # Populate related workshops
+      result[:related_workshops][fac_name] = related_workshop_ids
+
+      # Populate result[:facilitator_response_counts] =
+      # {this_workshop, all_my_workshops => {fac_id => {submission_type => count}}}
+      result[:facilitator_response_counts][:all_my_workshops][facilitator_id] =
+        {submission_type => submissions.count}
+      result[:facilitator_response_counts][:this_workshop][facilitator_id] =
+        {submission_type => submissions.where(pd_workshop_id: current_workshop_id).count}
+
+      # Populate result[:facilitator_averages]
+      # result[:facilitator_averages][fac_name][qname] = {this_workshop, all_my_workshops => score}
+      # result[:facilitator_averages][:questions] = {qname => qtext}
+      avg_scores = result[:facilitator_averages]
+      avg_scores[fac_name] = {}
+
+      qtexts = {}   # {qname => text}
+      summaries.each do |summary|
+        scope =
+          if summary[:workshop_id] == current_workshop_id
+            :this_workshop
+          elsif !summary[:workshop_id]
+            :all_my_workshops
+          else
+            :wrong_scope  # TODO: add non-fatal error
+          end
+        next if scope == :wrong_scope
+
+        q_name = summary[:name]
```

Comment:
another nit: `qname` feels more consistent with your other variable names, and that's how you refer to this variable in the comment on line 47. I would slightly prefer to write out the word "question" in all these variable names, but it's probably a matter of opinion whether that makes it easier to read the code in this case since it appears so often! :)

Reply:
will change to `qname`